File: packages/react-native-storybook/ios/ConfigHelper.m
#import "ConfigHelper.h"
#import "FileSystemHelper.h"

static NSString *const CONFIG_FILENAME = @"config.sherlo";
static NSString *const LOG_TAG = @"ConfigHelper";

/**
 * Helper for loading and managing Sherlo configuration.
 * Provides methods to read and parse configuration from the filesystem.
 */
@implementation ConfigHelper

/**
 * Loads and parses the Sherlo configuration file.
 * Attempts to decode the file content as base64 before parsing as JSON.
 * Returns an empty dictionary if the file doesn't exist or cannot be parsed.
 *
 * @param fileSystemHelper Helper class for file system operations
 * @return The parsed configuration as a NSDictionary
 */
+ (NSDictionary *)loadConfigWithFileSystemHelper:(FileSystemHelper *)fileSystemHelper {
    NSError *error = nil;
    NSString *configContent = [fileSystemHelper readFileAsString:CONFIG_FILENAME error:&error];
    
    if (error) {
        NSLog(@"[%@] Error reading config file: %@", LOG_TAG, error.localizedDescription);
        return nil;
    }
    
    if (!configContent || configContent.length == 0) {
        NSLog(@"[%@] Config file is empty", LOG_TAG);
        return nil;
    }
    
    // Try to decode if base64 encoded
    NSData *decodedData = nil;
    @try {
        decodedData = [[NSData alloc] initWithBase64EncodedString:configContent options:0];
        if (decodedData) {
            configContent = [[NSString alloc] initWithData:decodedData encoding:NSUTF8StringEncoding];
        }
    } @catch (NSException *exception) {
        NSLog(@"[%@] Config is not base64 encoded, trying to parse as plain JSON", LOG_TAG);
    }
    
    // Parse JSON
    NSError *jsonError = nil;
    NSData *jsonData = [configContent dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *config = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&jsonError];
    
    if (jsonError) {
        NSLog(@"[%@] Invalid JSON in config file: %@", LOG_TAG, jsonError.localizedDescription);
        return nil;
    }
    
    return config;
}

/**
 * Determines the initial application mode based on the configuration.
 * Checks for an override mode in the config first, otherwise defaults to testing mode 
 * if config exists, or default mode if there's no valid config.
 * 
 * @param config The configuration object
 * @return The determined mode (default, storybook, or testing)
 */
+ (NSString *)determineModeFromConfig:(NSDictionary *)config {
    @try {
        if (config && config.count > 0) {
            NSString *overrideMode = config[@"overrideMode"];

            if (overrideMode) {
                NSLog(@"[%@] Running in %@ mode", LOG_TAG, overrideMode);
                return overrideMode;
            }
            
            return MODE_TESTING;
        }
    } @catch (NSException *exception) {
        NSLog(@"[%@] Failed to determine initial mode: %@", LOG_TAG, exception.reason);
    }
    
    return MODE_DEFAULT;
}

@end

File: packages/react-native-storybook/android/src/main/java/io/sherlo/storybookreactnative/ConfigHelper.java
package io.sherlo.storybookreactnative;

import android.util.Base64;
import android.util.Log;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.File;

/**
 * Helper class for loading and managing Sherlo configuration.
 */
public class ConfigHelper {
    public static final String TAG = "ConfigHelper";
    private static final String CONFIG_FILENAME = "config.sherlo";

    private final FileSystemHelper fileSystemHelper;
    private final String syncDirectoryPath;

    public ConfigHelper(FileSystemHelper fileSystemHelper, String syncDirectoryPath) {
        this.fileSystemHelper = fileSystemHelper;
        this.syncDirectoryPath = syncDirectoryPath;
    }

    /**
     * Load and parse the Sherlo configuration file.
     * 
     * @return The parsed config as JSONObject, or empty JSONObject if no config exists
     * @throws Exception if there's an error reading or parsing the config
     */
    public JSONObject loadConfig() throws Exception {
        String configPath = this.syncDirectoryPath + "/" + CONFIG_FILENAME;
        Log.i(TAG, "Looking for config file at: " + configPath);

        // Check if config file exists
        File sherloConfigFile = new File(configPath);
        if (!sherloConfigFile.exists()) {
            Log.i(TAG, "Config file does not exist");
            return new JSONObject();
        }

        // Read config file
        String configJson = fileSystemHelper.readFile(configPath);
        if (configJson == null || configJson.trim().isEmpty()) {
            Log.w(TAG, "Config file is empty");
            return new JSONObject();
        }

        // Try to decode if base64 encoded
        try {
            byte[] decodedBytes = Base64.decode(configJson, Base64.DEFAULT);
            configJson = new String(decodedBytes, "UTF-8");
        } catch (Exception e) {
            Log.w(TAG, "Config is not base64 encoded, trying to parse as plain JSON");
        }

        // Parse JSON
        try {
            return new JSONObject(configJson);
        } catch (JSONException e) {
            Log.e(TAG, "Invalid JSON in config file: " + configJson);
            throw new Exception("Config file contains invalid JSON: " + e.getMessage());
        }
    }

    /**
     * Determines the initial mode based on configuration
     * 
     * @param config The config object
     * @return The determined mode (default, storybook, or testing)
     */
    public String determineInitialMode(JSONObject config) {
        try {
            if (config.length() > 0) {
                // Check for override mode
                String overrideMode = getOverrideMode(config);
                if (overrideMode != null) {
                    Log.i(TAG, "Running in " + overrideMode + " mode");
                    return overrideMode;
                }
                
                return SherloModuleCore.MODE_TESTING;
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to determine initial mode", e);
        }
        return SherloModuleCore.MODE_DEFAULT;
    }

    /**
     * Check if the config has an override mode specified.
     * 
     * @param config The config object
     * @return The override mode or null if not present
     */
    private String getOverrideMode(JSONObject config) {
        try {
            if (config.has("overrideMode")) {
                return config.getString("overrideMode");
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error reading overrideMode: " + e.getMessage());
        }
        return null;
    }
} 